### PR TITLE
Tests for main

### PR DESCRIPTION
**Pull request by NKH123**
Tests for main.go
Output:-
PASS
ok      github.com/SUMUKHA-PK/Raft-Distributed-Consensus/cmd    0.003s

### CODE REVIEW

File: cmd/main_test.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+		_, err := regexp.MatchString(`^(?:[0-9]{1,3}\.){3}[0-9]{1,3}$`, IP)
+		if err != nil {
+			t.Fatalf("Invalid IP Address in config file")
+		}
+		Port := v.Port
+		port, err := strconv.Atoi(Port)
+		if err != nil {
+			t.Fatalf("Invalid Port")
+		}
+		if port < 1024 {
```
Comment (SUMUKHA-PK) [inline cmd/main_test.go]: We dont need this, the server that I created will check for this.

File: cmd/main_test.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+func TestMain(t *testing.T) {
+	//check if the config file exists
+	info, err := os.Stat("../server.config.json")
+	if os.IsNotExist(err) {
+		t.Fatalf("Config file does not exist")
+	}
+	//check if there is no read permission to config file
+	mask := info.Mode()
+	if mask&(1<<2) == 0 {
```
Comment (SUMUKHA-PK) [inline cmd/main_test.go]: Is this necessary? What happens when the file is read protected and its opened? If theres no explicit error for this, include this test.

**Comment by SUMUKHA-PK** [comment]
@NKH123 , attach the output of this in the PR description.

File: cmd/main_test.go (hunk 2 above)
Comment (NKH123) [reply cmd/main_test.go]: You are checking if the read bit is on before opening. What else needs to be done?

Comment (SUMUKHA-PK) [reply cmd/main_test.go]: I'm telling we might not need an explicit test if the file opener already throws an error.

Check whether that happens, if it doesn't let it stay.

File: cmd/main_test.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	"strconv"
+	"testing"
+
+	"github.com/SUMUKHA-PK/Raft-Distributed-Consensus/types"
+)
+
+func TestMain(t *testing.T) {
+	//check if the config file exists
+	info, err := os.Stat("../server.config.json")
+	if os.IsNotExist(err) {
```
Comment (SUMUKHA-PK) [inline cmd/main_test.go]: In all places where theres an error, let following be the format.
`t.Fatalf("whatever you want to say : %v",err)

File: cmd/main_test.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+		t.Fatalf("Config file does not exist")
+	}
+	//check for Empty config file
+	if info.Size() == 0 {
+		t.Fatalf("Config file is empty")
```
Comment (SUMUKHA-PK) [inline cmd/main_test.go]: Same here

File: cmd/main_test.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	err = decoder.Decode(&configuration)
+	if err != nil {
+		t.Fatalf("Invalid config file")
+	}
+	//check for invalid IP and port
+	for _, v := range configuration {
+		IP := v.IP
+		_, err := regexp.MatchString(`^(?:[0-9]{1,3}\.){3}[0-9]{1,3}$`, IP)
+		if err != nil {
+			t.Fatalf("Invalid IP Address in config file")
```
Comment (SUMUKHA-PK) [inline cmd/main_test.go]: Here

File: cmd/main_test.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+		}
+		port := v.Port
+		if _, err := strconv.Atoi(port); err != nil {
+			t.Fatalf("Invalid Port Number in config file")
```
Comment (SUMUKHA-PK) [inline cmd/main_test.go]: Here

File: cmd/main_test.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+	//Check for opening of file
+	file, err := os.Open("../server.config.json")
+	if err != nil {
+		t.Fatalf("Error opening file")
+	}
+	decoder := json.NewDecoder(file)
+	//check for Invalid config file
+	configuration := make(map[int]types.Server)
+	err = decoder.Decode(&configuration)
+	if err != nil {
```
Comment (SUMUKHA-PK) [inline cmd/main_test.go:32]: Here

File: cmd/main_test.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+	if os.IsNotExist(err) {
+		t.Fatalf("Config file does not exist")
+	}
+	//check for Empty config file
+	if info.Size() == 0 {
+		t.Fatalf("Config file is empty")
+	}
+	//Check for opening of file
+	file, err := os.Open("../server.config.json")
+	if err != nil {
```
Comment (SUMUKHA-PK) [inline cmd/main_test.go:25]: Here

File: cmd/main_test.go (hunk 3 above)
Comment (SUMUKHA-PK) [inline cmd/main_test.go:16]: If there are errors in any place, let following be the format of printing.
`t.Fatalf("Whatever you want to say... : %v",err)`

**Pull request merged**